portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -28,33 +28,89 @@
 @WebServlet("/comments")
 public class DataServlet extends HttpServlet {
 
-  private final List<Comment> sampleComments = new ArrayList<Comment>();
+  private final String UNSAFE_CHARACTERS_REGEX = "[^A-Za-z0-9._~()'!*:@,;+?\\s-]";
 
-  @Override
-  public void init() {
-    // Add sample data
-    sampleComments.add(new Comment(
-      "Richie Goulazian",
-      "This is an example of a comment. This is one of multiple comments that I need to write"
-    ));
-    sampleComments.add(new Comment(
-      "Rich Goulaz",
-      "This is not an example of a comment... or maybe it is? How to tell..."
-    ));
-    sampleComments.add(new Comment(
-      "Bob Guy",
-      "My name is Bob. I like this page."
-    ));
-  }
+  private final List<Comment> sessionComments = new ArrayList<Comment>();
+  private final Gson gson = new Gson();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Convert comments to JSON
     Gson gson = new Gson();
-    String commentJson = gson.toJson(sampleComments);
+    String commentJson = gson.toJson(sessionComments);
 
     // Send the JSON as the response
     response.setContentType("application/json;");
     response.getWriter().println(commentJson);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get new comment parameters
+    final String COMMENT_AUTHOR_PARAM = "author";
+    final String COMMENT_BODY_PARAM = "comment-body";
+    String commentAuthor = getParameter(request, COMMENT_AUTHOR_PARAM, "");
+    String commentBody = getParameter(request, COMMENT_BODY_PARAM, "");
+
+    // Remove unsafe characters
+    commentAuthor = commentAuthor.replaceAll(UNSAFE_CHARACTERS_REGEX, "");
+    commentBody = commentBody.replaceAll(UNSAFE_CHARACTERS_REGEX, "");
+
+    // Validate the comment parameters
+    String validationError = validateIncomingComment(commentAuthor, commentBody);
+    if (validationError != null) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      response.setContentType("text/html;");
+      response.getWriter().println(validationError);
+      return;
+    }
+
+    // Add the comment to the session comments
+    Comment newComment = new Comment(commentAuthor, commentBody);
+    sessionComments.add(newComment);
+
+    String commentJson = gson.toJson(newComment);
+
+    // Return the new comment
+    response.setStatus(HttpServletResponse.SC_CREATED);
+    response.setContentType("application/json;");
+    response.getWriter().println(commentJson);
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * Validates the comment parameters of a new comment.
+   * @return Return the error string or null if there are no errors.
+   */
+  private String validateIncomingComment(String commentAuthor, String commentBody) {
+    String validationErrors = "";
+
+    // Check if author or body fields are blank
```

Comment:
This comment is redundant and should be removed because `commentAuthor.isBlank()` and  `commentBody.isBlank()` are self-explanatory.

Here's a snippet from an internal doc that you may or may not have access to (https://engdoc.corp.google.com/eng/doc/devguide/java/practices/comments.md?cl=head#what_for): 

What comments are for
Use a comment when it is infeasible to make your code self-explanatory. Here's a great quote from Robert C. Martin's Clean Code:

The proper use of comments is to compensate for our failure to express ourself in code. Note that I used the word failure. I meant it. Comments are always failures. We must have them because we cannot always figure out how to express ourselves without them, but their use is not a cause for celebration.... Every time you express yourself in code, you should pat yourself on the back. Every time you write a comment, you should grimace and feel the failure of your ability of expression.

Self-explanatory code beats liberally-commented code (which, of course, beats code that can't be understood). If you need a comment to explain what a block of code does, try to introduce an explaining variable or to extract a method. If you've already done that but you still need a comment, try to rename it until it answers all your concerns. Sometimes, just choosing better identifier names can be enough; sometimes it takes more work. It's almost always worth it.

Reply:
Solid quote and solid advice. I'll do a pass over all of the additions with this in mind.